XenModel/Actions/Updates/DownloadAndUpdateClientAction.cs

```diff
@@ -0,0 +1,292 @@
+                            address, errorCount, MAX_NUMBER_OF_TRIES);
+
+                        if (updateDownloadError == null)
+                            log.Error("An unknown error occurred.");
+                        else
+                            log.Error(updateDownloadError);
+                    }
+                } while (errorCount < MAX_NUMBER_OF_TRIES && needToRetry);
+            }
+            finally
+            {
+                //deregister download events
+                client.DownloadProgressChanged -= client_DownloadProgressChanged;
+                client.DownloadFileCompleted -= client_DownloadFileCompleted;
+
+                NetworkChange.NetworkAvailabilityChanged -= NetworkAvailabilityChanged;
+
+                client.Dispose();
+            }
+
+            //if this is still the case after having retried MAX_NUMBER_OF_TRIES number of times.
+            if (updateDownloadState == DownloadState.Error)
+            {
+                log.ErrorFormat("Giving up - Maximum number of retries ({0}) has been reached.", MAX_NUMBER_OF_TRIES);
+
+                MarkCompleted(updateDownloadError ?? new Exception(Messages.ERROR_UNKNOWN));
+            }
+
+        }
+
+        private void NetworkAvailabilityChanged(object sender, NetworkAvailabilityEventArgs e)
+        {
+            if (!e.IsAvailable && client != null && updateDownloadState == DownloadState.InProgress)
+            {
+                updateDownloadError = new WebException(Messages.NETWORK_CONNECTIVITY_ERROR);
+                updateDownloadState = DownloadState.Error;
+                client.CancelAsync();
+            }
+        }
+
+        protected override void Run()
+        {
+            if (downloadUpdate)
+            {
+                log.InfoFormat("Downloading '{0}' installer (from '{1}') to '{2}'", updateName, address, outputPathAndFileName);
+                Description = string.Format(Messages.DOWNLOAD_CLIENT_INSTALLER_ACTION_DESCRIPTION, updateName);
+                LogDescriptionChanges = false;
+                DownloadFile();
+                LogDescriptionChanges = true;
+
+                if (IsCompleted || Cancelled)
+                    return;
+
+                if (Cancelling)
+                    throw new CancelledException();
+            }
+            
+            ValidateMsi();
+
+            // Install the msi            
+            try
+            {
+                if (File.Exists(outputPathAndFileName))
+                {
+                    // Start the install process, it will handle closing of application.
+                    Process.Start(outputPathAndFileName);
+                    log.DebugFormat("Update {0} found and install started", updateName);
+                }
+            }
+            catch (Exception e)
+            {
+                if (File.Exists(outputPathAndFileName))
+                    File.Delete(outputPathAndFileName);

```

Comment:
I don't think that `File.Delete` raises an exception if the file had already been deleted.

Reply:
I dont see how it would as it only deletes if the file exists.

Reply:
@AaronRobson is right, but I think wherever this check occurs in XenCenter, it is there as an extra safety measure against an argument exception that Delete will throw if the parameter should ever be empty or null. Does not harm to keep it.